### PR TITLE
Fix bug 26449: Configurable product attribute issue

### LINKED ISSUE #26449: Configurable product attribute issue
### Preconditions (*)

1. Magento 2.3.1 with sample data
2. Size attribute Use in Layered Navigation

### Steps to reproduce (*)

1. Log into the admin account.
2. Navigate to Catalog > Products.
3. In add new product dropdown choose configurable product and create a configurable product using “Size” attribute:-

Screenshot_22

3. Size attribute available to choose in configurable, but leave it as it is.
Once you created the configurable product,  size attribute also assigned to main configurable product like this:
Screenshot_23

3. Once you filter the products on category page using 49 size, this product visible in filter result. As I am not using 49 size in configurable product. (49 size are not available in my store).

### Expected result (*)

1. Size attribute should not be assign to parent configurable product, so once user filter the product on category page using 49 size, this product not show in filter result.

### Actual result (*)

1.  49 Size assigned parent configurable product automatically and visible in filter result on category page when user select size 49. That is wrong functionality.

Let me now if you need more explanation to understand this issue.

### PR DESCRIPTION
**Pull request by tna274**
### Description (*)

Fix bug 26449: when creating a new configurable product, configuration options attribute of parent product still have value

### Related Pull Requests

N/A

### Fixed Issues (if relevant)

1. Fixes magento/magento2#26449: Configurable product attribute issue

### Manual testing scenarios (*)

1. Log into the admin account.
2. Navigate to Catalog > Products.
3. In add new product dropdown choose configurable product and create a configurable product using “Size” attribute:-
4. One of those variants value('Size') must be set default
Screenshot_22

image

3. After saving product, the value of "Size" attribute will be set null

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php [hunk 1]
```diff
@@ -51,6 +57,29 @@ public function beforeSave(
...
+    private function resetConfigurableOptionsData($object)
+    {
+        $extensionAttribute = $object->getExtensionAttributes();
+        if ($extensionAttribute && $extensionAttribute->getConfigurableProductOptions()) {
+            /** @var ProductAttributeRepositoryInterface $productAttributeRepository */
+            $productAttributeRepository = \Magento\Framework\App\ObjectManager::getInstance()
+                ->get(ProductAttributeRepositoryInterface::class);
+            /** @var OptionInterface $option */
+            foreach ($extensionAttribute->getConfigurableProductOptions() as $option) {
+                $eavAttribute = $productAttributeRepository->get($option->getAttributeId());
```
Comment (kandy) [inline app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php]: Loading attributes in the loop isn't best idea

Comment (tna274) [reply app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php]: Hi @kandy 
Could you suggest me an alternative way?

Comment (tna274) [reply app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php]: I've updated it

File: app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php [hunk 2]
```diff
@@ -21,18 +28,42 @@ class Product
...
      * @param ActionInterface $productIndexer
+     * @param ProductAttributeRepositoryInterface $productAttributeRepository
+     * @param \Magento\Framework\Api\SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param \Magento\Framework\Api\FilterBuilder $filterBuilder
      */
     public function __construct(
         Configurable $configurable,
-        ActionInterface $productIndexer
+        ActionInterface $productIndexer,
+        ProductAttributeRepositoryInterface $productAttributeRepository,
```
Comment (edenduong) [inline app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php]: Please use `ProductAttributeRepositoryInterface $productAttributeRepository = null` and `$this->productAttributeRepository = $productAttributeRepository ?: ObjectManager::getInstance()->get(ProductAttributeRepositoryInterface::class)` for backward compatible

**Review by edenduong — CHANGES_REQUESTED** [review]

File: app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php (hunk 2 above)
Comment (edenduong) [reply app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php]: Similar other classes

**Review by edenduong — CHANGES_REQUESTED** [review]
Hi @tna274 : Please check the Unit Test. It is failed now. Please run `vendor/bin/phpunit  app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php` to check the error of Unit Test. Thanks.

image

**Comment by tna274** [comment]
Hi @edenduong 
I have been updated code to fix the unit test

**Review by edenduong — CHANGES_REQUESTED** [review]
Method `$object->setData($optionAttribute->getAttributeCode(), null);` (and the logic in `if` block) is not covered by Unit Test. Please mockup`getExtensionAttributes` , to cover the `if` block.
Thank you.

File: app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php [hunk 3]
```diff
@@ -51,6 +87,39 @@ public function beforeSave(
...
+                ->setConditionType('in')
+                ->setValue($attributeIds)
+                ->create();
+            $this->searchCriteriaBuilder->addFilters([$filter]);
+            $searchCriteria = $this->searchCriteriaBuilder->create();
+            $optionAttributes = $this->productAttributeRepository->getList($searchCriteria)->getItems();
+
+            foreach ($optionAttributes as $optionAttribute) {
+                $object->setData($optionAttribute->getAttributeCode(), null);
+            }
```
Comment (edenduong) [inline app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php:124]: Should be covered by Unit Test

File: app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php [hunk 4]
```diff
@@ -61,6 +69,8 @@ public function testBeforeSaveConfigurable()
 
         $object->expects($this->once())->method('getTypeId')->will($this->returnValue(Configurable::TYPE_CODE));
         $object->expects($this->once())->method('getTypeInstance')->will($this->returnValue($type));
+        $object->expects($this->any())->method('getExtensionAttributes');
+        $object->expects($this->any())->method('setData');
```
Comment (edenduong) [inline app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php]: It has no meaning. It doesn't cover anything.

**Comment by tna274** [comment]
Hi @edenduong 
I have updated unit test

File: app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php [hunk 5]
```diff
@@ -31,18 +34,46 @@ class ProductTest extends \PHPUnit\Framework\TestCase
...
+     */
+    private $prdAttributeRepository;
+    /**
+     * @var SearchCriteriaBuilder|\PHPUnit\Framework\MockObject\MockObject
+     */
+    private $searchCriteriaBuilder;
+    /**
+     * @var FilterBuilder|\PHPUnit\Framework\MockObject\MockObject
+     */
+    private $filterBuilder;
```
Comment (edenduong) [inline app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php]: add `Mock` suffix to all mock variable

File: app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php [hunk 6]
```diff
@@ -31,18 +34,46 @@ class ProductTest extends \PHPUnit\Framework\TestCase
...
 
     public function setUp()
     {
         $this->objectManager = new \Magento\Framework\TestFramework\Unit\Helper\ObjectManager($this);
         $this->configurableMock = $this->createMock(Configurable::class);
         $this->actionMock = $this->createMock(ActionInterface::class);
+//        $this->prdAttributeRepository = $this->createMock(ProductAttributeRepositoryInterface::class);
```
Comment (edenduong) [inline app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php]: Remove redundant code

File: app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php [hunk 7]
```diff
@@ -21,18 +29,45 @@ class Product
...
+     * @param \Magento\Framework\Api\SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param \Magento\Framework\Api\FilterBuilder $filterBuilder
      */
     public function __construct(
         Configurable $configurable,
-        ActionInterface $productIndexer
+        ActionInterface $productIndexer,
+        ProductAttributeRepositoryInterface $productAttributeRepository = null,
+        \Magento\Framework\Api\SearchCriteriaBuilder $searchCriteriaBuilder = null,
+        \Magento\Framework\Api\FilterBuilder $filterBuilder = null
```
Comment (edenduong) [inline app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php]: Should `use`

File: app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php [hunk 8]
```diff
@@ -52,15 +83,58 @@ public function testBeforeSaveConfigurable()
...
             \Magento\ConfigurableProduct\Model\Product\Type\Configurable::class,
             ['getSetAttributes']
         );
+
+        $extensionAttributes = $this->createPartialMock(
+            \Magento\Framework\Api\ExtensionAttributesInterface::class,
+            ['getConfigurableProductOptions']
+        );
+        $option = $this->createPartialMock(
+            \Magento\ConfigurableProduct\Model\Product\Type\Configurable\Attribute::class,
```
Comment (edenduong) [inline app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php]: Please declare `use` class first

**Review by edenduong — CHANGES_REQUESTED** [review]
Hi @tna274 : Please check my comment
- Remove redundant code
- `use` instead of full class name
- Mock object should be declare as postfix `mock` like `searchCriteriaBuilderMock`

**Comment by tna274** [comment]
Hi @edenduong 
I have updated it

**Review by edenduong — DISMISSED** [review]
Thank you for your contribution !

**Comment by magento-engcom-team** [comment]
Hi @edenduong, thank you for the review.
[ENGCOM-7214](https://jira.corp.magento.com/browse/ENGCOM-7214) has been created to process this Pull Request

File: app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php [hunk 9]
```diff
@@ -31,18 +39,45 @@ class ProductTest extends \PHPUnit\Framework\TestCase
      * @var \Magento\ConfigurableProduct\Plugin\Model\ResourceModel\Product
      */
     private $model;
+    /**
```
Comment (eduard13) [inline app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php]: Add an empty line after each property definition

File: app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php [hunk 10]
```diff
@@ -31,18 +39,45 @@ class ProductTest extends \PHPUnit\Framework\TestCase
...
+     * @var FilterBuilder|\PHPUnit\Framework\MockObject\MockObject
+     */
+    private $filterBuilderMock;
 
     public function setUp()
     {
         $this->objectManager = new \Magento\Framework\TestFramework\Unit\Helper\ObjectManager($this);
         $this->configurableMock = $this->createMock(Configurable::class);
         $this->actionMock = $this->createMock(ActionInterface::class);
+        $this->prdAttributeRepositoryMock = $this->getMockBuilder(ProductAttributeRepositoryInterface::class)
```
Comment (eduard13) [inline app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php]: Please rename the property accordingly
```suggestion
        $this->productAttributeRepositoryMock = $this->getMockBuilder(ProductAttributeRepositoryInterface::class)
```

**Review by eduard13 — CHANGES_REQUESTED** [review]
Hi @tna274, thank you for your contribution, could you please review the bellow comments?
Thank you.

File: app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php [hunk 11]
```diff
@@ -9,6 +9,14 @@
 use Magento\Catalog\Model\Product\Type;
 use Magento\ConfigurableProduct\Model\Product\Type\Configurable;
 use Magento\Framework\Indexer\ActionInterface;
+use Magento\Catalog\Api\ProductAttributeRepositoryInterface;
+use Magento\Framework\Api\SearchCriteriaBuilder;
+use Magento\Framework\Api\SearchCriteria;
+use Magento\Framework\Api\FilterBuilder;
+use Magento\Framework\Api\ExtensionAttributesInterface;
+use Magento\ConfigurableProduct\Model\Product\Type\Configurable\Attribute as ConfigurableAttribute;
```
Comment (eduard13) [inline app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php]: Also sort these imports ascending.

**Comment by tna274** [comment]
Hi @eduard13 
I have updated my pull request

**Comment by tna274** [comment]
Hi @eduard13 
Could you check my PR again please?

File: app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php [hunk 12]
```diff
@@ -4,11 +4,21 @@
...
 namespace Magento\ConfigurableProduct\Plugin\Model\ResourceModel;
 
+use Magento\Catalog\Api\ProductAttributeRepositoryInterface;
 use Magento\ConfigurableProduct\Model\Product\Type\Configurable;
 use Magento\Framework\Indexer\ActionInterface;
+use Magento\ConfigurableProduct\Api\Data\OptionInterface;
+use Magento\Catalog\Api\Data\ProductAttributeInterface;
+use Magento\Framework\App\ObjectManager;
+use Magento\Framework\Api\SearchCriteriaBuilder;
+use Magento\Framework\Api\FilterBuilder;
```
Comment (edenduong) [inline app/code/Magento/ConfigurableProduct/Plugin/Model/ResourceModel/Product.php]: Pleaes use Ctrl + Alt + O in Phpstorm to sort the order of import

**Review by edenduong — CHANGES_REQUESTED** [review]
Please check my comment. Please use Ctrl + Alt + O in Phpstorm to optimize the order of import.

File: app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php (hunk 11 above)
Comment (edenduong) [reply app/code/Magento/ConfigurableProduct/Test/Unit/Plugin/Model/ResourceModel/ProductTest.php]: @tna274 : You can use Ctrl + Alt + O in Phpstorm to Optimize Import (sort the order)

**Comment by tna274** [comment]
Hi @edenduong 
Thanks for your comment. I have updated my PR.

**Review by edenduong — APPROVED** [review]
@tna274 : Thank you. I think it is ok now.

**Comment by magento-engcom-team** [comment]
Hi @edenduong, thank you for the review.
[ENGCOM-7214](https://jira.corp.magento.com/browse/ENGCOM-7214) has been created to process this Pull Request

**Comment by eduard13** [comment]
@magento run all tests

**Comment by engcom-Delta** [comment]
@magento run Functional Tests CE, Functional Tests EE, Database Compare

**Comment by engcom-Delta** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @engcom-Delta. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @engcom-Delta, here is your [new Magento instance](https://pr-27339.instances.magento-community.engineering).
Admin access: [https://pr-27339.instances.magento-community.engineering/admin_5282](https://pr-27339.instances.magento-community.engineering/admin_5282)
Login: `8444a943` Password: `a969e450925a`
Instance will be terminated in up to 3 hours.

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed according **Manual testing scenarios**

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B, Functional Tests CE

**Comment by engcom-Echo** [comment]
@magento run Functional Tests CE

**Review by slavvka — CHANGES_REQUESTED** [review]
@tna274 please resolve the conflicts and re-run tests after upgrade to PHP 7.4

**Comment by engcom-Echo** [comment]
I will take care of test coverage MFTF.

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-7214](https://jira.corp.magento.com/browse/ENGCOM-7214) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Go to Admin->Catalog->Products->Add Configurable Product;
2. Fill all required fields and click on **Create Configurations->Create New Attribute**;
3. Create text swatch attribute "**Size**" and set value **s** -> **Is Default**;
Screenshot from 2020-05-19 16-40-04
Screenshot from 2020-05-19 16-40-29
4. Save Attribute and go to the next step;
5. Select all options expect **s** (default);
Screenshot from 2020-05-19 16-41-56
6. **Generate products** and **Save product**
Screenshot from 2020-05-19 16-44-35
7. Go to Storefront -> category with created before product;

**Before:**
:heavy_multiplication_x: **S** Size assigned parent configurable product automatically and visible in filter result on category page when user select size **S**. That is the wrong functionality.

Peek 2020-05-19 16-55

**After:**
:heavy_check_mark: Size attribute assigned to parent configurable product, so once user filter the product on a category page using **S** size, this product does not show in filter result.
screenshot_16

**Pull request merged**